Initial workflow

devngc opened the pull request:
This resolves #1

mostaphaRoudsari commented on honeybee_vtk/helper.py:
```diff
@@ -0,0 +1,147 @@
+        Example of structure will be [[X, Y, Z], [X, Y, Z], [X, Y, Z]].
+    """
+    # Create Ladybug Point3D objects from boundary points
+    lb_boundary_pts = [
+        Point3D(point[0], point[1], point[2])
+        for point in boundary_points]
```
It is a style thing but I honestly think keeping it in a single line is easier to read. Also you can use * to unwrap the input arguments.

```suggestion
    lb_boundary_pts = [Point3D(*point) for point in boundary_points]
```

mostaphaRoudsari commented on honeybee_vtk/helper.py:
```diff
@@ -0,0 +1,147 @@
+        A boolean value.
+    """
+
+    lb_boundary_pts = [
+        Point3D(point[0], point[1], point[2]) for point in boundary_points]
```
I'm not going to comment on all the instances of code like this. Refer to the one earlier.

mostaphaRoudsari commented on honeybee_vtk/helper.py:
```diff
@@ -0,0 +1,147 @@
+    return face.center, face.normal
+
+
+def get_face_center(points_lst):
```
I have a few comments about this function.

1. names like `points_lst` makes me cringe. let's use annotations. points: List[List[Tuple]]]
2. it looks like you want to change `face_center` function to `_face_center` and then rename this function to `face_center`. Right now having `face_center` and `get_face_center` can be really confusing.
3. finally, this function doesn't return face_center as far as I can see!

mostaphaRoudsari commented on honeybee_vtk/helper.py:
```diff
@@ -0,0 +1,147 @@
+        
+        - normals: A list of Ladybug Vector3D objects.
+    """
+    start_points = [face_center(points)[0] for points in points_lst]
+    normals = [face_center(points)[1] for points in points_lst]
+    end_points = [
+        get_end_point(start_points[i], normals[i]) for i in range(len(start_points))]
```
You are rebuilding the face twice for each face. What if you do it once? I assume this function will be called several times during the translation and this will become a performance issue that we can avoid.

mostaphaRoudsari commented on honeybee_vtk/tovtk.py:
```diff
@@ -0,0 +1,311 @@
+"""Functions to create vtk objects based on the data extracted from hbjson."""
```
I'm a fan of making the module names human readable. I would use `to_vtk` instead of `tovtk`

mostaphaRoudsari commented on honeybee_vtk/tovtk.py:
```diff
@@ -0,0 +1,311 @@
+
+    Args:
+        points_lst: A list of lists of points for the geometry in the HBJSON. Here,
+            each point is a list of X, Y, and Z cordinates of the point.
```
```suggestion
            each point is a list of X, Y, and Z coordinates of the point.
```

mostaphaRoudsari commented on honeybee_vtk/tovtk.py:
```diff
@@ -0,0 +1,311 @@
+    """
+    grouped_points = {face_type: [] for face_type in hb_types}
+
+    for i in range(len(hb_types)):
+        grouped_points[hb_types[i]].append(points_lst[i])
```
Can you provide a small sample here. I'm very confused on how this is supposed to work. Probably a test can also be very helpful.

devngc replied:
Example provided

mostaphaRoudsari commented on honeybee_vtk/tovtk.py:
```diff
@@ -0,0 +1,311 @@
+    polygon with holes.
+
+    Args:
+        points: A list of points. Here, each point is a list of X, Y, and Z cordinate.
```
Find and replace `cordinate` with `coordinate`

mostaphaRoudsari commented on honeybee_vtk/tovtk.py:
```diff
@@ -0,0 +1,311 @@
+        vtk_points.InsertNextPoint(tuple(point))
+
+    for i in range(len(points)):
+        vtk_polygon.GetPointIds().SetId(i, i)
```
This is strange to me that the id is a tuple but I guess that's how VTK works.

mostaphaRoudsari commented on honeybee_vtk/tovtk.py:
```diff
@@ -0,0 +1,311 @@
+            and Z cordinate.
+
+    Returns:
+        A vtk object that with multiple polygons.
```
```suggestion
        A vtk object with multiple polygons.
```

mostaphaRoudsari commented on honeybee_vtk/tovtk.py:
```diff
@@ -0,0 +1,311 @@
+    """
+    vtk_polydata_lst = []
+
+    for i in range(len(points_lst)):
+
+        vtk_points, vtk_polygon = create_polygon(points_lst[i])
```
```suggestion
    for points in points_lst:

        vtk_points, vtk_polygon = create_polygon(points )
```

mostaphaRoudsari commented on honeybee_vtk/tovtk.py:
```diff
@@ -0,0 +1,311 @@
+
+    vtk_polydata_extended = vtk.vtkAppendPolyData()
+    
+    for i in range(len(vtk_polydata_lst)):
```
Almost every time that you write `range(len(some_list))` there is probably an easier way to do it in Python including here.

mostaphaRoudsari commented on honeybee_vtk/tovtk.py:
```diff
@@ -0,0 +1,311 @@
+        file_name: A text string for the the file name to be written.
+    """
+    vtk_polydata_extended = create_polygons(polydata)
+    writer = vtk.vtkPolyDataWriter()
```
Isn't this the legacy writer? Why don't we use the XML one?

devngc replied:
Yes. legacy VTK files are smaller in size than VTP which is an XML file.

mostaphaRoudsari replied:
Can we add an argument on the top to make this and option so user can decide if the file should be exported as a legacy or an XML file?

devngc replied:
Done

mostaphaRoudsari commented on honeybee_vtk/tovtk.py:
```diff
@@ -0,0 +1,311 @@
+    writer = vtk.vtkPolyDataWriter()
+    writer.SetFileName(file_name + '.vtk')
+    writer.SetInputData(point_polydata)
+    writer.Write()
```
Maybe it will be nice to return the file path from these write functions.

devngc replied:
Okay. Can you please share what can we do with those file paths?

mostaphaRoudsari replied:
It will be easier to use. Let's say we are developing a command line tool that uses this function and we want to access the file after its being written. It will be handy to get the filepath from this function instead of trying to create it. It will also ensure changing the path/name will not break any downstream code.

devngc replied:
Done

mostaphaRoudsari commented on honeybee_vtk/tovtk.py:
```diff
@@ -0,0 +1,311 @@
+    writer.Write()
+
+
+def write_arrows(start_points, end_points, normals, file_name):
```
This method is probably much easier to use if we pick two of these 3 inputs and calculate the other one inside the function itself. Assuming that I understand it correctly the normal is the vector that connects start and end points. Also in that case we should probably call them vectors and not normals.

mostaphaRoudsari commented on honeybee_vtk/tovtk.py:
```diff
@@ -0,0 +1,311 @@
+    lines_polydata = []
+    cones_polydata = []
+
+    for i in range(len(start_points)):
```
The pythonic approach is to use `zip`!

mostaphaRoudsari commented on honeybee_vtk/writer.py:
```diff
@@ -0,0 +1,141 @@
+        file_path, *, file_name=None, get_grids=True,
+        get_vectors=True):
+
+    """Reads a valid HBJSON and writes a .zip of vtk files to the disk.
```
See 4th item here: https://www.python.org/dev/peps/pep-0257/#one-line-docstrings

```suggestion
    """Read a HBJSON and write a .zip of vtk files.
```

mostaphaRoudsari commented on honeybee_vtk/writer.py:
```diff
@@ -0,0 +1,141 @@
+        file_name: A text string for the name of the .zip file to be written. If no
+            text string is provided, the name of the HBJSON file will be used as a
+            file name for the .zip file.
+        get_grids: A boolean. Defaults to True. Grids will not be extracted from HBJSON
```
We usually use `include` and `exclude` in these cases. E.g. include_grids

mostaphaRoudsari commented on honeybee_vtk/writer.py:
```diff
@@ -0,0 +1,141 @@
+            set to False.
+    """
+    # HBJSON file
+    if file_path:
```
`file_path` is a required argument. You don't need this check.

mostaphaRoudsari commented on honeybee_vtk/writer.py:
```diff
@@ -0,0 +1,141 @@
+        raise ValueError(
+            'Not a valid JSON file.'
+            )
+    else:
```
so many if/else clauses here. makes it hard to read the code. I think you can just continue the code here without using else.

mostaphaRoudsari commented on honeybee_vtk/writer.py:
```diff
@@ -0,0 +1,141 @@
+                file_names.append('grid base')
+
+            # If base_geometry is not found but mesh faces are found
+            if grids[1]:
```
shouldn't this be `elif`?

devngc replied:
No. grids is a tuple of three items.

mostaphaRoudsari replied:
but you want to only generate the 'grid mesh' if there is no 'grid base'. right? Or do you want to add these filenames anyways?

devngc replied:
It's not just about the file names. Check_grid function returns a tuple of three items. A list of sensorgirds with base_geometry, A list of sensorgrids with mesh and a list of sensorgrids with points. We need to check all three of them and write to file if they are found.

mostaphaRoudsari replied:
Thank you. I think that's the part that I didn't understand. Not knowing about the overall workflow, I think I would change  the code to write the first one that it finds and ignore the rest.

So if grid-base is provided it means we can keep it simple and only write that and save time for generating the rest. If it is not provided then we check the next option and so on but you know the overall process better so I leave it to you.

devngc replied:
We need to check all three as it often is the case where a Sensorgrid object in HBJSON has different keys. An HBJSON can have all three types of sensorgrids;
1. That has `base_geometry` and `mesh` as keys
2. That has only `mesh` as key
3. That has neither `base_geometry` nor `mesh` as keys
I have already seen files having two cases from above. And I can see that it's very easy to generate an HBJSON that will have all the three above case. That is why I designed this behaviour. I can share some HBJSON files to demo this.

mostaphaRoudsari commented on honeybee_vtk/writer.py:
```diff
@@ -0,0 +1,141 @@
+                file_names.append('grid mesh')
+
+            # If only grid points are found
+            if grids[2]:
```
```suggestion
            elif grids[2]:
```

devngc replied:
No. grids is a tuple of three items.

mostaphaRoudsari commented on honeybee_vtk/writer.py:
```diff
@@ -0,0 +1,141 @@
+                start_points, end_points, vectors = get_grid_points(grids[2])
+                write_color_grouped_points(start_points, vectors)
+                file_names.append('grid points')
+        else:
+            pass
```
why do you need this?

mostaphaRoudsari commented on honeybee_vtk/writer.py:
```diff
@@ -0,0 +1,141 @@
+        # Write vectors if they are requested
+        if get_vectors:
+            
+            # If Aperture objects are there in HBJSON
+            if 'Aperture' in hb_types:
+                # Create face normals
+                start_points, end_points, normals = get_face_center(
+                    grouped_points['Aperture'])
+                write_arrows(start_points, end_points, normals, 'Aperture')
+                file_names.append('Aperture vectors')
+            
+            # If grids are found in HBJSON
+            if grids:
+                
+                # If base_geometry is found in any of the grids
+                if grids[0]:
+                    base_geo_points = get_grid_base(grids[0])[0]
+                    start_points, end_points, normals = get_face_center(base_geo_points)
+                    write_arrows(start_points, end_points, normals, 'grid base')
+                    file_names.append('grid base vectors')
+
+                # If mesh is found in any of the grids
+                if grids[1]:
+                    mesh_points = get_grid_mesh(grids[1])[0]
+                    start_points, end_points, normals = get_face_center(mesh_points)
+                    write_arrows(start_points, end_points, normals, 'grid mesh')
+                    file_names.append('grid mesh vectors')
+        else:
+            pass
```
I would move each of these `if` clauses to a separate private function in this module.

if get_vectors:
    _get_vectors(..., )

mostaphaRoudsari commented on honeybee_vtk/writer.py:
```diff
@@ -0,0 +1,141 @@
+                os.remove(file_name + '.vtk')
+            except OSError:
+                warnings.warn(
+                    'Honeybee does not have permission to delete files. You can delete' 
```
I would print the name of the exact vtk file here. It's hard to know which file should be removed.

mostaphaRoudsari reviewed (commented):
Hi @devngc, thank you! see my comments. We should add at least one test here that does the translation.

devngc commented on .gitignore:
```diff
@@ -0,0 +1,22 @@
+/.vscode
+.eggs
+*.code-workspace
+*.vtk
+*.zip
+*.glance
```
Will only need .zip that I added back. Will add .glance later when I have figured out the way to generate it.

devngc replied:
I realized we don't need to catalogue .glance or .zip since those files are really disposable.

devngc commented on deploy.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+
+echo "Building distribution"
+python setup.py sdist bdist_wheel
+echo "Pushing new version to PyPi"
+twine upload dist/* -u $PYPI_USERNAME -p $PYPI_PASSWORD
```
I don't understand really how this works.

mostaphaRoudsari replied:
It's all good. You can disregard this comment for now.

devngc commented on honeybee_vtk/hbjson.py:
```diff
@@ -0,0 +1,231 @@
+
+    points_lst, hb_type_lst = get_data(hbjson['orphaned_faces'],
+        'face_type')
+    points.extend(points_lst)
+    hb_types.extend(hb_type_lst)
```
This empty list is hosts other lists of point3ds. Please let me know if this is unacceptable.

mostaphaRoudsari replied:
I found it strange. You can use the return lists from the first function.

```python
points, hb_type = get_data(hbjson['orphaned_faces'], face_type)
```

and then extend the list as you go.

devngc replied:
I am confused. Is it not what I am doing [here](https://github.com/ladybug-tools/honeybee-vtk/pull/3/files/ce41ea95bf6f0eb795b76179ce4559bd9d20210d#diff-413c4f0c714d3ef8eaee2fc67a3f051ade252e4ec342c2ab6d14e72ea958b32aR84)?

mostaphaRoudsari replied:
No but it is not important. Both of them works.

devngc commented on honeybee_vtk/hbjson.py:
```diff
@@ -0,0 +1,231 @@
+    points = []
+    hb_types = []
+
+    points_lst, hb_type_lst = get_data(hbjson['orphaned_faces'],
```
That will only happen if there only room objects in HBJSON and no other face. I believe it's unlikely to have such a model with only boxes.

mostaphaRoudsari replied:
What about single room energy models? They usually will not have `orphaned_faces`.

devngc replied:
I am adding support for this now.

devngc replied:
Done

devngc commented on honeybee_vtk/hbjson.py:
```diff
@@ -0,0 +1,231 @@
+        for grid in hbjson['properties']['radiance']['sensor_grids']:
+            if 'base_geometry' in grid:
+                grid_with_base.append(grid)
+            elif 'mesh' in grid and 'base_geometry' not in grid:
```
And that's what we want. Whenever base_geometry is there, we want to use it rather than the meshes which will mean more polygons.

mostaphaRoudsari replied:
Agreed. That means the second check is redundant. If the code is here we already know that `base_geometry` is not in grid. So you don't have to check for it.

devngc commented on honeybee_vtk/hbjson.py:
```diff
@@ -0,0 +1,231 @@
+    vectors = []
+
+    for grid in grids:
+        for face in grid['base_geometry']:
```
This function only gets used if there's base_geometry in a Sensorgrid object. [This](https://github.com/ladybug-tools/honeybee-vtk/pull/3/files#diff-413c4f0c714d3ef8eaee2fc67a3f051ade252e4ec342c2ab6d14e72ea958b32aR104) function checks that.

mostaphaRoudsari replied:
Makes sense. Thanks.

devngc commented on honeybee_vtk/hbjson.py:
```diff
@@ -0,0 +1,231 @@
+from .helper import get_point3d, get_end_point
+
+
+def get_data(hbjson_obj, hb_type):
```
I added annotations but kept the argument names. I feel it's easy to understand the argument type with these names.

mostaphaRoudsari replied:
Can you point me to where you added these annotations? I can't seem to find them.

mostaphaRoudsari replied:
Isn't HBJSON object from type Dict? Up to you but all this `var_lst` and `var_obj` looks ugly to me. The `obj` one is also confusing because everything in Python is technically an object so you are not really adding useful information here.

devngc replied:
I revised this. Also, I renamed all the variables with "lst" yesterday.

devngc commented on honeybee_vtk/hbjson.py:
```diff
@@ -0,0 +1,231 @@
+    points = []
+    hb_types = []
+
+    points_lst, hb_type_lst = get_data(hbjson['orphaned_faces'],
+        'face_type')
```
I fixed these. I am using 89 characters long line length.

mostaphaRoudsari reviewed (approved):
I added my main comments and suggestions. The code works and now we have tests. Let's merge this in and I will send a PR to set-up the actions which you can review. That will make the version 0.1.0 available to everyone.

devngc commented:
Thank you @mostaphaRoudsari for your comments. I believe I have addressed all your comments today except a couple of them for which I responded with an explanation.
